### PR TITLE
Forward-port $collection->sortBy() rewrite from v2

### PR DESCRIPTION
**Pull request by lukasbestle**
I have improved the v2 version of my `$collection->sortBy()` rewrite with some further fixes and have also adapted it to the v3 code style.

The tests run through fine (I have added some edge cases as well), but we might need to test it a bit more manually before merging.

### CODE REVIEW

**Comment by lukasbestle** [comment]
I used `ContentField` in the test as I needed an object with a `__toString()` method to test sorting of objects. I don't specifically need that exact class though. :)

I don't think we should move this to the CMS Collection tests as the test is relevant for any kind of object (so should be part of the Toolkit tests). But you are right, I will find a different solution. Wait a moment.

**Comment by distantnative** [comment]
I see 👍 In that case, I'd suggest to define a mock object class with `__toString()` in the current test file, instead of going for an actual class from the Cms namespace.

File: src/Collection/Traits/Sorter.php [hunk 1]
```diff
@@ -17,37 +18,64 @@
...
+
+        // loop through all method arguments and find sets of fields to sort by
+        $fields = [];
+        foreach ($args as $arg) {
+            // get the index of the latest field array inside the $fields array
+            $currentField = ($fields)? count($fields) - 1 : 0;
+
+            // detect the type of argument
+            // sorting direction
+            $argLower = Str::lower($arg);
```
Comment (bastianallgeier) [inline src/Collection/Traits/Sorter.php]: Do we really need the Str::lower method here?

File: src/Collection/Traits/Sorter.php [hunk 2]
```diff
@@ -17,37 +18,64 @@
...
-                ${"param_$i"} = $args[$i];
+                $fields[$currentField]['flags'] = $arg;
             }
-            $params[] = &${"param_$i"};
         }
 
+        // build the multisort params in the right order
+        $params = [];
+        foreach ($fields as $field) {
+            $params[] = A::get($field, 'values',    []);
```
Comment (bastianallgeier) [inline src/Collection/Traits/Sorter.php]: I would replace all the A::get method calls with the ?? operator.

File: src/Collection/Traits/Sorter.php (hunk 1 above)
Comment (lukasbestle) [reply src/Collection/Traits/Sorter.php]: I think so. Otherwise it wouldn't work if you passed something like `ASC` or `Asc` as the sorting direction. Please keep in mind that the lowercase argument variant is only ever being used for the sorting direction, other types of arguments use the original argument.

Comment (bastianallgeier) [reply src/Collection/Traits/Sorter.php]: No, I meant that we could just use strtolower here, because we don't expect any mb_string worthy stuff anyway.

Comment (bastianallgeier) [reply src/Collection/Traits/Sorter.php]: Just trying to stay with native PHP functions as often as possible.

File: src/Collection/Traits/Sorter.php (hunk 2 above)
Comment (lukasbestle) [reply src/Collection/Traits/Sorter.php]: Sure, haven't thought of that. Done! :)

File: src/Collection/Traits/Sorter.php (hunk 1 above)
Comment (lukasbestle) [reply src/Collection/Traits/Sorter.php]: Ah, makes sense. Done!
